### PR TITLE
Added an example on how to deploy all the edge components

### PR DESCRIPTION
**Pull request by e-minguez**

### CODE REVIEW

**Comment by e-minguez** [comment]
Neuvector missing

File: examples/edge-stack-iso/eib.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      version: 0.6.3
+    repositories:
+    - name: jetstack
+      url: https://charts.jetstack.io
+    - name: longhorn
+      url: https://charts.longhorn.io
+    - name: rancher-stable
+      url: https://releases.rancher.com/server-charts/stable
+    - name: suse-edge
+      url: https://suse-edge.github.io/charts
```
Comment (e-minguez) [inline examples/edge-stack-iso/eib.yaml:45]: https://github.com/suse-edge/edge-image-builder/issues/306 will simplify this a lot

File: examples/edge-stack-iso/kubernetes/helm/values/rancher.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+hostname: rancher-192.168.122.11.sslip.io
```
Comment (e-minguez) [inline examples/edge-stack-iso/kubernetes/helm/values/rancher.yaml:1]: This requires the nginx ingress configuration tweaks

File: examples/edge-stack-iso/kubernetes/manifests/hello-node.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+apiVersion: apps/v1
```
Comment (e-minguez) [inline examples/edge-stack-iso/kubernetes/manifests/hello-node.yaml:1]: This is just an example of a workload being deployed

File: examples/edge-stack-iso/custom/files/metal3.sh [hunk 4]
```diff
@@ -0,0 +1,128 @@
+#!/bin/bash
```
Comment (e-minguez) [inline examples/edge-stack-iso/custom/files/metal3.sh:1]: We need to find a better way to do this 😅

File: examples/edge-stack-iso/kubernetes/helm/values/neuvector.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# Based on https://gist.github.com/agracey/6966338d34c7277f5b5ec6f4c1431b1b
```
Comment (e-minguez) [inline examples/edge-stack-iso/kubernetes/helm/values/neuvector.yaml:1]: I'm using your example here @agracey is this right?

**Comment by jdob** [comment]
One quick request: can you add an entry under the Advanced section of the README.md in the examples directory? There is a lot going on in this example (which is awesome) but mere mortals aren't going to have a clue what all is in it.

**Comment by e-minguez** [comment]
> One quick request: can you add an entry under the Advanced section of the README.md in the examples directory? There is a lot going on in this example (which is awesome) but mere mortals aren't going to have a clue what all is in it.

Yes but no :D 

I would like to first discuss the approach before writing the docs just in case it changes a lot.

File: examples/edge-stack-iso/kubernetes/manifests/neuvector-namespace.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    pod-security.kubernetes.io/enforce: privileged
```
Comment (e-minguez) [inline examples/edge-stack-iso/kubernetes/manifests/neuvector-namespace.yaml:5]: https://github.com/neuvector/neuvector-helm?tab=readme-ov-file#deploy-in-kubernetes

File: examples/edge-stack-iso/custom/files/metal3.sh [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+      export GOARCH="amd64" ;;
+    "aarch64")
+      export GOARCH="arm64" ;;
+    "*")
+      echo "Arch not found, asumming amd64"
+      export GOARCH="amd64" ;;
+  esac
+
+  # Clusterctl bin
+  # Maybe just use the binary from hauler if available
```
Comment (atanasdinov) [inline examples/edge-stack-iso/custom/files/metal3.sh:73]: I'd suggest dropping this comment. If we want to sideload the binary, we should instead just put it under `/custom/files`.

**Review by atanasdinov — COMMENTED** [review]
Looks great!

Some minor comments but we definitely need to add docs for this before merging as @jdob pointed out.

**Comment by e-minguez** [comment]
> Looks great!
> 
> Some minor comments but we definitely need to add docs for this before merging as @jdob pointed out.

I'll take a look at the comments, thanks.

WRT docs I agree but I want to get all the inputs first then clean all the reviews/suggestions before just in case it changes too much.

File: examples/edge-stack-iso/custom/files/basic-setup.sh [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+export K3SKUBECTL="/opt/bin/kubectl"
+export RKE2KUBECONFIG="/etc/rancher/rke2/rke2.yaml"
+export K3SKUBECONFIG="/etc/rancher/k3s/k3s.yaml"
+
+export SUSECAFILE="/usr/share/pki/trust/anchors/SUSE_Trust_Root.crt.pem"
+export SUSECACM="suse-internal-ca"
+export SUSECACMNAMESPACE="kube-system"
+
+########################
+# METAL3 CHART DETAILS #
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/custom/files/basic-setup.sh:13]: I'll get rid of most of them. These are leftovers from the automation I currently use.

File: examples/edge-stack-iso/custom/files/edge-stack-setup.service [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ConditionPathExists=/opt/edge/bin/basic-setup.sh
+
+[Service]
+User=root
+Type=forking
+# Metal3 can take A LOT to download the IPA image
+TimeoutStartSec=1800
+
+ExecStartPre=/bin/sh -c "echo 'Starting to setup Edge components'"
+# Scripts are executed in StartPre because it can be multiple ones
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/custom/files/edge-stack-setup.service:16]: It can be just a single 'ExecStart' per systemd service file, so instead of having multiple services, the scripts are executed as 'pretasks' then the start is just an echo.

File: examples/edge-stack-iso/custom/files/metal3.sh [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  exit 0
+fi
+
+# Wait for metal3
+while ! ${KUBECTL} wait --for condition=ready -n ${METAL3_CHART_TARGETNAMESPACE} $(${KUBECTL} get pods -n ${METAL3_CHART_TARGETNAMESPACE} -l app.kubernetes.io/name=metal3-ironic -o name) --timeout=10s; do sleep 2 ; done
+
+# Get the ironic IP
+IRONICIP=$(${KUBECTL} get cm -n ${METAL3_CHART_TARGETNAMESPACE} ironic-bmo -o jsonpath='{.data.IRONIC_IP}')
+
+# This only works for metallb, we need to do the same for nodeport I guess
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/custom/files/metal3.sh:25]: This script is intended to be used when deploying an HA cluster as it deploys metallb by default and an IP will be used to expose the Ironic API (https://github.com/suse-edge/charts/tree/main/charts/metal3/0.6.3#metallb-optional) . If there is just a single node, there is no metallb (unless explicitly deployed) so this won't work and instead we can use NodePort https://github.com/suse-edge/charts/blob/main/charts/metal3/0.6.3/charts/ironic/values.yaml#L93

File: examples/edge-stack-iso/custom/files/metal3.sh [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+# Wait for metallb
+while ! ${KUBECTL} wait --for condition=ready -n ${METALLBNAMESPACE} $(${KUBECTL} get pods -n ${METALLBNAMESPACE} -l app.kubernetes.io/component=controller -o name) --timeout=10s; do sleep 2 ; done
+
+# Don't create the ippool if already created
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/custom/files/metal3.sh:30]: I feel it is clean to avoid applying the thing if it is already there.

File: examples/edge-stack-iso/custom/files/metal3.sh (hunk 7 above)
Comment (e-minguez) [reply examples/edge-stack-iso/custom/files/metal3.sh:73]: As hauler allows you to pack binaries (https://rancherfederal.github.io/hauler-docs/docs/guides-references/hauler-content/files) as well I think it would be cleaner if we can add a flow to EIB to save binaries to hauler somehow.

```
hauler store add file https://github.com/kubernetes-sigs/cluster-api/releases/download/v${METAL3_CLUSTERCTLVERSION}/clusterctl-linux-${GOARCH}
```

I can put the binary directly I agree but I don't want to do save it directly into git and if I don't save it the example will not be 'complete'.

File: examples/edge-stack-iso/custom/files/rancher.sh [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+fi
+
+# Get or create the lock to run all those steps just in a single node
+# As the first node is created WAY before the others, this should be enough
+# TODO: Investigate if leases is better
+if [ $(${KUBECTL} get cm -n ${RANCHERLOCKNAMESPACE} ${RANCHERLOCKCMNAME} -o name | wc -l) -lt 1 ]; then
+  ${KUBECTL} create configmap ${RANCHERLOCKCMNAME} -n ${RANCHERLOCKNAMESPACE} --from-literal foo=bar
+else
+  exit 0
+fi
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/custom/files/rancher.sh:22]: The point of this 'lock' is because this script will be executed on all the hosts, so it is a poor's man 'distributed/shared lock' :D If a file is being written by one of the hosts, the others won't be able to see it (unless I'm missing something)

File: examples/edge-stack-iso/custom/scripts/99_edge-setup.sh [hunk 13]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+# Copy the basic setup script from combustion to the final location
+mkdir -p /opt/edge/bin/
+cp basic-setup.sh /opt/edge/bin/
+chmod a+x /opt/edge/bin/basic-setup.sh
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/custom/scripts/99_edge-setup.sh:6]: You are right

File: examples/edge-stack-iso/eib.yaml [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    apiVIP: 192.168.122.10
+  nodes:
+  - hostname: host1rke2
+    initializer: true
+    type: server
+  - hostname: host2rke2
+    type: server
+  - hostname: host3rke2
+    type: server
+  version: v1.27.10+rke2r1
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/eib.yaml:66]: Yep https://www.suse.com/suse-rancher/support-matrix/all-supported-versions/rancher-v2-8-2/

File: examples/edge-stack-iso/custom/files/rancher.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (atanasdinov) [reply examples/edge-stack-iso/custom/files/rancher.sh:22]: This is precisely the case. Isn't Rancher supposed to be bootstrapped just once?

File: examples/edge-stack-iso/custom/files/edge-stack-setup.service (hunk 9 above)
[the remark this answers is not in the document]
Comment (atanasdinov) [reply examples/edge-stack-iso/custom/files/edge-stack-setup.service:16]: Can we reword this to `Scripts are executed in StartPre because Start can only run a single one`?

File: examples/edge-stack-iso/eib.yaml [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+  isoConfiguration:
+    installDevice: /dev/vda
+    unattended: true
+  packages:
+    additionalRepos:
+    - url: http://download.suse.de/ibs/SUSE:/CA/SLE_15/
+    packageList:
+    - jq
+    - ca-certificates-suse
+    - qemu-guest-agent
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/eib.yaml:77]: For the hypervisor to be able to 'talk' with the VM. I think it doesn't hurt and there are some benefits (keeping the time in sync, the ability to freeze the VM when taking snapshots, etc.) https://qemu-project.gitlab.io/qemu/interop/qemu-ga.html

File: examples/edge-stack-iso/custom/files/basic-setup.sh [hunk 16]
```diff
@@ -0,0 +1,117 @@
+#!/bin/bash
+# Pre-requisites. Cluster already running
+export RKE2KUBECTL="/var/lib/rancher/rke2/bin/kubectl"
+export K3SKUBECTL="/opt/bin/kubectl"
```
[the remark this answers is not in the document]
Comment (atanasdinov) [reply examples/edge-stack-iso/custom/files/basic-setup.sh:4]: Yes but we're providing specific examples here and not an automation that lots of customers will need. I believe being straight to the point with each example and only including the relevant bits and pieces for it is the best but I'd let others chime in on this.

File: examples/edge-stack-iso/custom/files/metal3.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (atanasdinov) [reply examples/edge-stack-iso/custom/files/metal3.sh:25]: Sure but given this is an example for an HA cluster then this comment is not necessary. Perhaps relevant to my point here: https://github.com/suse-edge/edge-image-builder/pull/331#discussion_r1537343681

[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/custom/files/metal3.sh:25]: I'll remove the comment

File: examples/edge-stack-iso/custom/files/rancher.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge-stack-iso/custom/files/rancher.sh:22]: I'm sorry but I feel I'm missing something. The script will be executed on all the hosts being provisioned as there is no way to make a distinction right now on EIB. If there is no lock it can happen that all the hosts are waiting for rancher to be up (next line) and then they try to create the password, etc. at the same time and then the script may fail on those that are not the first one.
If I create a file locally, it will be created on each host, leading to the previous scenario as well as there is no 'sync' between them in that case.

File: examples/edge-stack-iso/custom/files/basic-setup.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (hardys) [reply examples/edge-stack-iso/custom/files/basic-setup.sh:4]: We only expect to support RKE2 for the CAPI/Metal3 case (management cluster and downstream clusters) so I would agree it may be simpler and less confusing to remove the k3s options for now?

File: examples/edge-stack-iso/custom/files/metal3.sh (hunk 7 above)
Comment (atanasdinov) [reply examples/edge-stack-iso/custom/files/metal3.sh:71]: I see. Keeping the download is completely fine for now. On a separate note, Hauler is intentionally not being used for files right now and this might very well be the case in the future as well but it's something we'll discuss when we have the time.

File: examples/edge-stack-iso/kubernetes/helm/values/neuvector.yaml (hunk 5 above)
Comment (e-minguez) [reply examples/edge-stack-iso/kubernetes/helm/values/neuvector.yaml:1]: Better use this https://github.com/suse-edge/charts/blob/main/charts/neuvector-core/2.6.6/values.yaml
